### PR TITLE
LibGfx/LibVideo: Read batches of multiple bytes in VPX BooleanDecoder

### PR DESCRIPTION
**Pull request by Zaggy1024**
This does a few things:

- The decoder uses a 32- or 64-bit integer as a reservoir of the data being decoded, rather than one single byte as it was previously.
- `read_bool()` only refills the reservoir (value) when the size drops below one byte. Previously, it would read out a bit-sized range from the data to completely refill the 8-bit value, doing much more work than necessary for each individual read.
- Errors are now deferred until `finish_decode()` is finished, meaning branches to return errors only need to occur at the end of a ranged decode.
- VP9-specific code for reading the marker bit was moved to its own function in Context.h.
- A debug flag `VPX_DEBUG` was added to optionally enable checking of the final bits in a VPX ranged arithmetic decode and ensure that it contains all zeroes. These zeroes are a bitstream requirement for VP9, and are also present for all WebP test inputs we currently have. Therefore, it can be used to test whether all the data present in the range has been consumed, allowing us to more fully determine that the data has been read correctly.

A lot of the size of this diff comes from the removal of error handling from all the range decoder reads in LibVideo/VP9 and LibGfx/WebP (VP8), since it is now checked only at the end of the range.

This gives us about a 7% performance boost in my benchmark, but this difference becomes more apparent with further optimizations of other parts of the decoding process.

This should cause no behavioral changes.

### CODE REVIEW

**Comment by Zaggy1024** [comment]
cc @nico

**Comment by nico** [comment]
Very cool!

On my machine, with https://files.edge.network/misc/webp/file_example_WEBP_1500kB.webp , this speeds up

```sh
% time Build/lagom/image ~/Downloads/file_example_WEBP_1500kB.webp --no-output
```

from 0.377s to 0.29s , a ~23% speedup [1]


Do you have a sense of how much perf we leave on the table if we keep the `ErrorOr<u8>` (or u64 or whatever) interface but do all the other changes in here?



1: Details:

```sh
% git checkout zaggy^
% Meta/serenity.sh build lagom image TestImageDecoder file
% ~/src/hack/bench.py -o before.txt Build/lagom/image ~/Downloads/file_example_WEBP_1500kB.webp --no-output
% git checkout zaggy
% Meta/serenity.sh build lagom image TestImageDecoder file
% ~/src/hack/bench.py -o after.txt Build/lagom/image ~/Downloads/file_example_WEBP_1500kB.webp --no-output
% ministat before.txt after.txt 
x before.txt
+ after.txt
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|    + + + +                       +                                                                                                                                                  x x   x         x|
||_______M___A____________|                                                                                                                                                           |_____M_A______| |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    N           Min           Max        Median           Avg        Stddev
x   5     0.3742609    0.38220406    0.37754869    0.37826223  0.0036845294
+   5    0.28845787    0.30297804    0.29029894    0.29245119  0.0059803455
Difference at 95.0% confidence
	-0.085811 +/- 0.00724395
	-22.6856% +/- 1.91506%
	(Student's t, pooled s = 0.0049669)
```

**Comment by Zaggy1024** [comment]
> Do you have a sense of how much perf we leave on the table if we keep the ErrorOr<u8> (or u64 or whatever) interface but do all the other changes in here?

I've split off the change for deferring errors to a separate commit and added perf numbers. Looks like it's close to 5% improvement!

**Comment by Zaggy1024** [comment]
Had a mishap with my placement of the `finish_decode()` call in the VP8 decoder that made me think the padding zeroes weren't present for the second partition, but we should be good now.

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 1]
```diff
@@ -1221,6 +1219,9 @@ ErrorOr<void> decode_VP8_image_data(Gfx::Bitmap& bitmap, FrameHeader const& head
         }
     }
 
+    for (auto& decoder : streams)
+        TRY(decoder.finish_decode());
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:1222]: why add calls to this in the webp path? the vp8 spec doesn't say anything about the content of padding afaict.

(also, seems unrelated to the rest of the commit)

File: Userland/Libraries/LibGfx/ImageFormats/BooleanDecoder.cpp [hunk 2]
```diff
@@ -82,23 +85,26 @@ ErrorOr<bool> BooleanDecoder::read_bool(u8 probability)
...
-        return_value = (2 * return_value) + TRY(read_bool(128));
+        return_value = (2 * return_value) + read_bool(128);
     }
     return return_value;
 }
 
 ErrorOr<void> BooleanDecoder::finish_decode()
 {
+    if (m_overread)
+        return Error::from_string_literal("Range decoder was read past the end of its data");
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/BooleanDecoder.cpp:106]: I guess that's why you did the change in the first commit. I'd add a separate `did_read_too_many_bits()` or similar thingy for this, probably (and call it fro the `finish_decode_vp9()` function, I suppose).

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp (hunk 1 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:1222]: As noted in the definition of `finish_decode()`, the padding in our WebP test inputs also fill the remaining data in the range with zeroes, so this can be used to verify that the data is consumed correctly. That check is only enabled when VPX_DEBUG is enabled. Given that it seems consistent, I see no reason to separate this functionality off from the general BooleanDecoder implementation.

The function is also used in the second commit to check for deferred overread errors, but both parts of the function are of benefit.

Comment (Zaggy1024) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:1222]: If desired, the padding check can be made into a debug print instead of an error, though.

**Comment by nico** [comment]
The first commit looks excellent.

Having 5% more perf from the 2nd commit seems nice I suppose, but the `TRY` pattern is much nicer API-wise. We probably have way more than 5% of perf left, so personally I'd keep the `TRY()`s for now. But in the end, up to you and maintainers.

**Comment by Zaggy1024** [comment]
> The first commit looks excellent.
> 
> Having 5% more perf from the 2nd commit seems nice I suppose, but the `TRY` pattern is much nicer API-wise. We probably have way more than 5% of perf left, so personally I'd keep the `TRY()`s for now. But in the end, up to you and maintainers.

I agree the TRY pattern is nice, but I think that keeping it with a proven performance penalty isn't worth it. I'd rather have it be optional, which I've implemented as part of the VPX_DEBUG flag. If one wants to prevent further execution of the decode when an overread occurs, a `VERIFY()` can be placed after that as well.

The performance benefit should be entirely separate from most other optimizations we'll be able to do in these decoders, so reducing the time spent in arithmetic decoding, the more apparent the effects of those other optimizations will be, as well.

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp (hunk 1 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:1222]: Going to resolve this, but let me know if you would like the optional padding check to be a print instead of an error.
